runtime/onert/backend/train/optimizer/Optimizers.test.cc

```diff
@@ -0,0 +1,672 @@
+  AdamOptimizerVerifier &operator=(const AdamOptimizerVerifier &) = delete;
+
+public:
+  void verify()
+  {
+    for (uint32_t step = 0; step < _nums_step; ++step)
+    {
+      const T beta1_power = std::pow(_beta1, step + 1);
+      const T beta2_power = std::pow(_beta2, step + 1);
+
+      calculateExpected(beta1_power, beta2_power);
+
+      backend::train::optimizer::Adam::UpdateFactors factors{_gradient, _trainable, step};
+      _adam.applyGradient(factors);
+
+      const auto vars = _trainable.optVars();
+      const auto &m = *vars[0];
+      const auto &v = *vars[1];
+
+      auto actual_lr = _adam.getLearningRate(step);
+      auto expected_lr = [&]() {
+        auto biasCorrection = [&](double f) { return 1.0f - std::pow(f, step + 1); };
+        return _learning_rate * (std::sqrt(biasCorrection(_beta2)) / biasCorrection(_beta1));
+      }();
+      EXPECT_EQ(actual_lr, expected_lr);
+
+      for (size_t i = 0; i < _expected_trainable.size(); ++i)
+      {
+        EXPECT_NEAR(reinterpret_cast<T *>(m.buffer())[i], _expected_m[i], 1e-07f);
+        EXPECT_NEAR(reinterpret_cast<T *>(v.buffer())[i], _expected_v[i], 1e-07f);
+        EXPECT_NEAR(reinterpret_cast<T *>(_gradient.buffer())[i], _expected_gradient[i], 1e-07f);
+        EXPECT_NEAR(reinterpret_cast<T *>(_trainable.buffer())[i], _expected_trainable[i], 1e-07f);
+      }
+    }
+  }
+
+private:
+  void calculateExpected(const float beta1_power, const float beta2_power)
+  {
+    EXPECT_TRUE(_expected_trainable.size() == _expected_gradient.size());
+    EXPECT_TRUE(_expected_trainable.size() == _expected_m.size());
+    EXPECT_TRUE(_expected_trainable.size() == _expected_v.size());
+
+    const T alpha = _learning_rate * std::sqrt(static_cast<T>(1) - beta2_power) /
+                    (static_cast<T>(1) - beta1_power);
+
+    for (size_t i = 0; i < _expected_trainable.size(); ++i)
+    {
+      T &m = _expected_m[i];
+      T &v = _expected_v[i];
+      T &g = _expected_gradient[i];
+      T &t = _expected_trainable[i];
+
+      if (_use_nesterov)
+      {
+        m += (g - m) * (static_cast<T>(1) - _beta1);
+        v += (std::pow(g, 2) - v) * (static_cast<T>(1) - _beta2);
+        t -= ((g * (static_cast<T>(1) - _beta1) + _beta1 * m) * alpha) / (std::sqrt(v) + _epsilon);
+      }
+      else
+      {
+        m += (g - m) * (static_cast<T>(1) - _beta1);
+        v += (std::pow(g, 2) - v) * (static_cast<T>(1) - _beta2);
+        t -= (m * alpha) / (std::sqrt(v) + _epsilon);
+      }
+    }
+  }
+
+private:
+  backend::train::optimizer::Adam _adam;
+  MockUpTrainableTensor &_trainable;
+  const MockUpTensor &_gradient;
+  float _learning_rate;
+  float _beta1;
+  float _beta2;
+  float _epsilon;
+  bool _use_nesterov;
+  uint32_t _nums_step;
+
+  std::vector<T> _expected_trainable;
+  std::vector<T> _expected_gradient;
+  std::vector<T> _expected_m;
+  std::vector<T> _expected_v;
+};
+
+} // namespace
+
+TEST(Optimizer, SGDValueValidation)
+{
+  // Single step
```

Comment:
(Q) Why is this test single step? The training_steps is 10, isn't it?

Reply:
It's my mistake. I will fix it.